### PR TITLE
fix(python): accept more types in groupby.agg

### PR DESCRIPTION
**Pull request by kngwyu**
Just a minor detail, but showing e.g., `got Sequnce[str]` might be more useful than `expected 'Expr | Sequence[Expr]', got 'list'`.

### CODE REVIEW

**Comment by stinodego** [comment]
I think we should actually remove this check entirely, and simply defer to `selection_to_pyexpr_list(aggs)` for handling the input.

`df.groupby('a').agg('b')` should be perfectly fine input, but is currently not accepted due to this check.

**Comment by kngwyu** [comment]
@stinodego 
I see, but since `selection_to_pyexpr_list` doesn't do runtime type checking, I'm a bit worried about it can lead to lengthy rust errors shown. Anyway, I will reconsider this PR to allow more types handled properly.

**Comment by kngwyu** [comment]
>  I'm a bit worried that it can lead to lengthy rust errors shown. 

This was my misunderstanding and I think errors look OK since they are handled by `expr_to_lit_or_expr`.
So I'm going to remove this entire check, and in that case, what type should be used? There is `IntoExpr` alias, but it looks like broader types are accepted, as shown below.  https://github.com/pola-rs/polars/blob/29fd4c8bddf959abc18fcc4da375556965f8ed63/py-polars/polars/internals/expr/expr.py#L81-L86

**Comment by kngwyu** [comment]
I'm now leaning towards extending `IntoExpr` to `IntoExpr: TypeAlias = "PolarsExprType | PythonLiteral | pli.Series | None"` and using `agg: IntoExpr | Iterable[IntoExpr]` as an argument type of LazyFrame/DataFrame's agg. However, one problem is that we can't deny dicts (e.g., `{"b": "sum"}`), which are now considered as an example of a bad input in tests. Should we allow dict or restrict the input class (e.g., `IntoExpr | Sequence[IntoExpr]` might work) ?
Any idea is welcome.

**Comment by stinodego** [comment]
On bad examples in tests, you can simply use a `# type: ignore` comment. These are supposed to fail the type checker.

Regarding the input type, for now, you could do `str | PolarsExprType | Iterable[str | PolarsExprType]`

**Comment by kngwyu** [comment]
> On bad examples in tests, you can simply use a # type: ignore comment. These are supposed to fail the type checker.

Sorry, I meant that dicts are currently explicitly rejected since they are not instances of `Sequence` in the if sentence I'm removing in this PR. Maybe I would need another bad example.

>  Technically, literals work too but they don't make much sense in a groupby context.

That makes sense, but I feel like that could confuse users who might want to know what types can be convertible to Expr. I prefer prioritizing correctness in type annotations and recommending typical use cases in examples.

**Comment by kngwyu** [comment]
I renewed this PR to extend the arguments of `LazyFrame.agg` as my comment above.

**Comment by stinodego** [comment]
I really like what you've done here!

It looks like this check was introduced to solve some specific syntax confusion - but I think a streamlined and consistent API is more important than catching possibly unexpected behaviour for new Polars users. As long as we document the intended use properly.

`IntoExpr` is quite explicit - it tells me I can input _"anything that can be converted into an expression"_. Feels Rust-y too. And this alias makes it easier to type annotate similar methods.

Test cases look fine to me. We do now allow dicts, which is think is fine (see reasoning above).

I have one comment about using `Sequence` vs `Iterable`. Let me know your thoughts - otherwise I think this looks good.

@alexander-beedie would you mind taking a look too?

File: py-polars/polars/internals/dataframe/groupby.py
Comment (stinodego) [inline py-polars/polars/internals/dataframe/groupby.py:210]: As you noted, we accept iterables too. So in accordance with your statement to "\prefer prioritizing correctness in type annotations and recommending typical use cases in examples.", we should use `Iterable` here.

```suggestion
    def agg(self, aggs: IntoExpr | Iterable[IntoExpr]) -> pli.DataFrame:
```

**Review by stinodego — CHANGES_REQUESTED** [review]
Just the one Sequence -> Iterable change, and this has my approval.

**Comment by alexander-beedie** [comment]
I wouldn't hide the None in `IntoExpr` - that would be better if made explicit/visible in the signature; otherwise, I like the idea (and it looks good) ;)

File: py-polars/polars/internals/dataframe/groupby.py
Comment (kngwyu) [reply py-polars/polars/internals/dataframe/groupby.py]: Thanks, I first tried `IntoExpr | Sequence` and changed it to `Iterable`, so this left unchanged from the previous trial.

**Review by stinodego — APPROVED** [review]
Great, thanks!

**Pull request merged**
